### PR TITLE
KT-23137 Intention for converting to block comment and vise versa

### PR DESCRIPTION
**Pull request by t-kameyama**
https://youtrack.jetbrains.com/issue/KT-23137

### CODE REVIEW

File: idea/src/org/jetbrains/kotlin/idea/intentions/ConvertBlockCommentToLineCommentIntention.kt [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+package org.jetbrains.kotlin.idea.intentions
+
+import com.intellij.openapi.editor.Editor
+import com.intellij.psi.PsiComment
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiWhiteSpace
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.KtPsiFactory
+
+class ConvertBlockCommentToLineCommentIntention : SelfTargetingIntention<PsiComment>(
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/intentions/ConvertBlockCommentToLineCommentIntention.kt]: Should be probably not applicable to documentation comments:

```
/**
 * ...
 */
```

File: idea/src/org/jetbrains/kotlin/idea/intentions/ConvertBlockCommentToLineCommentIntention.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    override fun isApplicableTo(element: PsiComment, caretOffset: Int): Boolean {
+        return element.isBlockComment()
+    }
+
+    override fun applyTo(element: PsiComment, editor: Editor?) {
+        val psiFactory = KtPsiFactory(element)
+
+        val prevSibling = element.prevSibling
+        val indent = if (prevSibling is PsiWhiteSpace) {
+            val space = prevSibling.text.reversed().takeWhile { it == ' ' }
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/intentions/ConvertBlockCommentToLineCommentIntention.kt]: What about `'\t'`? Of course it's rare case but it's possible

File: idea/src/org/jetbrains/kotlin/idea/intentions/ConvertBlockCommentToLineCommentIntention.kt (hunk 1 above)
Comment (mglukhikh) [reply idea/src/org/jetbrains/kotlin/idea/intentions/ConvertBlockCommentToLineCommentIntention.kt]: it's enough to check that beginning of comment is `/**`

File: idea/src/org/jetbrains/kotlin/idea/intentions/ConvertLineCommentToBlockCommentIntention.kt [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        return element.isEndOfLineComment()
+    }
+
+    override fun applyTo(element: PsiComment, editor: Editor?) {
+        var firstComment = element
+        while (true) {
+            firstComment = firstComment.prevComment() ?: break
+        }
+
+        val indent = (firstComment.prevSibling as? PsiWhiteSpace)?.text?.reversed()?.takeWhile { it == ' ' } ?: ""
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/intentions/ConvertLineCommentToBlockCommentIntention.kt]: Here tabulation is also possible

File: idea/src/org/jetbrains/kotlin/idea/intentions/ConvertBlockCommentToLineCommentIntention.kt (hunk 2 above)
Comment (mglukhikh) [reply idea/src/org/jetbrains/kotlin/idea/intentions/ConvertBlockCommentToLineCommentIntention.kt]: Please check how this and other intention work with situation like below (multiple new-lines between comment and declaration).

```
/* 
   block comment 
*/


fun foo() {}
```

File: idea/src/org/jetbrains/kotlin/idea/intentions/ConvertBlockCommentToLineCommentIntention.kt (hunk 1 above)
Comment (t-kameyama) [reply idea/src/org/jetbrains/kotlin/idea/intentions/ConvertBlockCommentToLineCommentIntention.kt]: I added `docComment.kt` test. `isBlockComment()` function returns false, if element is KDoc.

**Comment by t-kameyama** [comment]
Fixed

**Comment by mglukhikh** [comment]
Thank you, merged.

**Pull request closed without merging**